plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/LibvirtStorageAdaptor.java

```diff
@@ -637,6 +638,38 @@ public KVMPhysicalDisk getPhysicalDisk(String volumeUuid, KVMStoragePool pool) {
         }
     }
 
+    /**
+     * Thread-safe increment storage pool usage refcount
+     * @param uuid UUID of the storage pool to increment the count
+     */
+    private void incStoragePoolRefCount(String uuid) {
+        synchronized (storagePoolRefCounts) {
+            int refCount = storagePoolRefCounts.computeIfAbsent(uuid, k -> 0);
+            refCount += 1;
+            storagePoolRefCounts.put(uuid, refCount);
+        }
+    }
+
+    /**
+     * Thread-safe decrement storage pool usage refcount for the given uuid and return if storage pool still in use.
+     * @param uuid UUID of the storage pool to decrement the count
+     * @return true if the storage pool is still used, else false.
+     */
+    private boolean decStoragePoolRefCount(String uuid) {
+        synchronized (storagePoolRefCounts) {
+            Integer refCount = storagePoolRefCounts.get(uuid);
+            if (refCount != null && refCount > 1) {
+                s_logger.debug(String.format("Storage pool %s still in use, refCount %d", uuid, refCount));
+                refCount -= 1;
+                storagePoolRefCounts.put(uuid, refCount);
+                return true;
+            } else {
+                storagePoolRefCounts.remove(uuid);
+                return false;
+            }
+        }
+    }
```

Comment:
@rp- , should inc and dec maybe both call the same bit of synchronised code so that no inc and dec can happen at the same time as well? The `storagePoolRefCounts` can be a synchronised map to prevent strangeties happening on the map, but we also want to prevent multiple access on the elements.

```suggestion
    /**
     * adjust refcount
     */
    private int adjustStoragePoolRefCount(Sting uuid, int adjustment) {
        synchronised(uuid) {
        // some access on the storagePoolRefCounts.key(uuid) element
            Integer refCount = storagePoolRefCounts.computeIfAbsent(uuid, k -> 0);
            refCount += adjustment;
            storagePoolRefCounts.put(uuid, refCount);
            if (refCount < 1) {
                storagePoolRefCounts.remove(uuid);
            } else {
                storagePoolRefCounts.put(uuid, refCount);
            }
        }
        return refCount;
    }
    /**
     * Thread-safe increment storage pool usage refcount
     * @param uuid UUID of the storage pool to increment the count
     */
    private void incStoragePoolRefCount(String uuid) {
        adjustStoragePoolRefCount(uuid, 1);
    }

    /**
     * Thread-safe decrement storage pool usage refcount for the given uuid and return if storage pool still in use.
     * @param uuid UUID of the storage pool to decrement the count
     * @return true if the storage pool is still used, else false.
     */
    private boolean decStoragePoolRefCount(String uuid) {
        return adjustStoragePoolRefCount(uuid, -1) > 0;
    }
```

???

Reply:
I can't fully follow you here, any map access is `synchronized` so there can't be any concurrent access to elements?

Reply:
Or it just stroke me, that maybe the `LibVirtStorageAdaptor` is no singleton object and there could be more instances, but than it would probably be good enough to make the map `static`?

Reply:
two remarks:
1. Can you considder a threadsafe collection ? i.e. concurrent map.
2. the map can be guarded (as concurrent map or by your own synchronised blocks) but I don't think that will guard its elements from being guarded, only the structure of the map itself.
3. (there is always a third) blocking access to the map globally may impact performance (though in this case that would have to be on very big hosts.

I think you are right about making the map static, btw.

Reply:
1. Sure I can also use concurrenthashmap
2. Sure if you leak the elements out of the synchronization blocks they are not thread safe anymore, but the current code doesn't do that.
3. As far as I read, concurrenthashmap only locks on element level, so switching to that would already be good.